File: web/src/upstreamAddr.ts
/**
 * 与后端 `internal/api/upstream_addr.go` 一致：仅接受 IPv4 或 IPv6（可带方括号）。
 * 返回 null 表示不合法；合法时返回 trim 后的地址（IPv6 无方括号，与入库一致由服务端再规范化）。
 */
export function normalizeUpstreamAddr(input: string): string | null {
  const t = input.trim();
  if (!t) return null;
  if (t.startsWith("[")) {
    const end = t.lastIndexOf("]");
    if (end <= 1) return null;
    const inner = t.slice(1, end);
    if (!inner.includes(":")) return null;
    try {
      new URL(`http://[${inner}]`);
      return inner;
    } catch {
      return null;
    }
  }
  if (!t.includes(":")) {
    return isIPv4DottedDecimal(t) ? t : null;
  }
  try {
    new URL(`http://[${t}]`);
    return t;
  } catch {
    return null;
  }
}

function isIPv4DottedDecimal(s: string): boolean {
  const parts = s.split(".");
  if (parts.length !== 4) return false;
  for (const p of parts) {
    if (!/^\d{1,3}$/.test(p)) return false;
    const n = Number(p);
    if (n < 0 || n > 255 || String(n) !== p) return false;
  }
  return true;
}

export function isUpstreamIpV4OrV6(input: string): boolean {
  return normalizeUpstreamAddr(input) !== null;
}

/** 上游协议字面量；与后端 internal/api/upstream_addr.go 常量保持同步。 */
export type UpstreamProtocol = "udp" | "dot" | "doh";

/** 协议对应的常用默认端口。 */
export function defaultPortForProtocol(p: UpstreamProtocol): number {
  switch (p) {
    case "dot":
      return 853;
    case "doh":
      return 443;
    default:
      return 53;
  }
}

/** 协议对应的中文展示名（表单与列表共用）。 */
export function protocolLabel(p: UpstreamProtocol | string): string {
  switch (p) {
    case "dot":
      return "DoT";
    case "doh":
      return "DoH";
    case "udp":
    default:
      return "UDP";
  }
}

/**
 * 归一 DoH 端点路径：
 * - 空 / "/" 视为缺省 "/dns-query"。
 * - 必须以 "/" 开头，且不含查询字符串或片段。
 * - 不合法返回 null。
 */
export function normalizeDoHPath(input: string): string | null {
  const t = input.trim();
  if (!t || t === "/") return "/dns-query";
  if (!t.startsWith("/")) return null;
  if (t.includes("?") || t.includes("#")) return null;
  return t;
}

/**
 * 归一 TLS SNI（可选）：空字符串视为不填，返回 ""；不合法返回 null。
 * 校验规则：1-253 字符；每段 1-63 字符；字母/数字/连字符；连字符不在首尾。
 * 也允许 IP 字面量直接做 SNI（与后端一致）。
 */
export function normalizeTLSServerName(input: string): string | null {
  const t = input.trim();
  if (!t) return "";
  if (t.length > 253) return null;
  if (isUpstreamIpV4OrV6(t)) return normalizeUpstreamAddr(t);
  const labels = t.replace(/\.$/, "").split(".");
  for (const label of labels) {
    if (!label || label.length > 63) return null;
    if (label.startsWith("-") || label.endsWith("-")) return null;
    if (!/^[A-Za-z0-9-]+$/.test(label)) return null;
  }
  return t;
}

File: internal/api/upstream_addr.go
package api

import (
	"net"
	"strings"
)

// 上游协议常量；与 DB CHECK 约束、forward 层分发保持同步。
const (
	UpstreamProtocolUDP = "udp"
	UpstreamProtocolDoT = "dot"
	UpstreamProtocolDoH = "doh"
)

// NormalizeUpstreamProtocol 大小写无关地归一协议字符串；空值视为 udp。
// 返回 (规范化值, 是否合法)。
func NormalizeUpstreamProtocol(s string) (string, bool) {
	s = strings.ToLower(strings.TrimSpace(s))
	if s == "" {
		return UpstreamProtocolUDP, true
	}
	switch s {
	case UpstreamProtocolUDP, UpstreamProtocolDoT, UpstreamProtocolDoH:
		return s, true
	}
	return "", false
}

// NormalizeUpstreamAddr 校验上游地址须为合法 IPv4 或 IPv6（不接受主机名）。
// 允许输入带方括号的 IPv6，如 [2001:db8::1]；返回规范字符串（与 net.IP.String() 一致，IPv6 无方括号）。
// 三种协议都共享此校验，避免引入 bootstrap DNS 的"鸡生蛋"问题。
func NormalizeUpstreamAddr(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", false
	}
	if len(s) >= 2 && s[0] == '[' {
		i := strings.LastIndex(s, "]")
		if i <= 1 {
			return "", false
		}
		inner := s[1:i]
		ip := net.ParseIP(inner)
		if ip == nil {
			return "", false
		}
		return ip.String(), true
	}
	ip := net.ParseIP(s)
	if ip == nil {
		return "", false
	}
	return ip.String(), true
}

// NormalizeDoHPath 归一 DoH 端点路径：
//   - 空 / "/" 视为缺省 "/dns-query"。
//   - 必须以 "/" 开头，不含查询字符串或片段。
//   - 返回 (规范化路径, 是否合法)。
func NormalizeDoHPath(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if s == "" || s == "/" {
		return "/dns-query", true
	}
	if !strings.HasPrefix(s, "/") {
		return "", false
	}
	if strings.ContainsAny(s, "?#") {
		return "", false
	}
	return s, true
}

// NormalizeTLSServerName 校验可选 SNI；空字符串视为缺省（保持 nil 由调用方处理）。
// 仅做 RFC1123 风格的简单校验：每段 1-63 字符，由字母/数字/连字符组成，连字符不在首尾。
func NormalizeTLSServerName(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", true
	}
	if len(s) > 253 {
		return "", false
	}
	// 允许 IP 字面量直接做 SNI（部分服务端用 IP SAN 证书）。
	if ip := net.ParseIP(s); ip != nil {
		return ip.String(), true
	}
	for _, label := range strings.Split(strings.TrimSuffix(s, "."), ".") {
		if label == "" || len(label) > 63 {
			return "", false
		}
		if label[0] == '-' || label[len(label)-1] == '-' {
			return "", false
		}
		for i := 0; i < len(label); i++ {
			c := label[i]
			ok := (c >= 'a' && c <= 'z') ||
				(c >= 'A' && c <= 'Z') ||
				(c >= '0' && c <= '9') ||
				c == '-'
			if !ok {
				return "", false
			}
		}
	}
	return s, true
}

// DefaultPortForProtocol 返回协议的常用默认端口。
func DefaultPortForProtocol(proto string) int {
	switch proto {
	case UpstreamProtocolDoT:
		return 853
	case UpstreamProtocolDoH:
		return 443
	default:
		return 53
	}
}
